### PR TITLE
Update percentage-wise confusion matrix

### PR DESCRIPTION
**Pull request by ATriantafyllopoulos**
Changes: Generated table now shows number of samples and percentages in parentheses.

@hagenw as discussed internally, this PR changes what confusion matrix with `percentage=True` returns. Background: I argued that the support information is critical for confusion matrices and was missing for percentage-wise ones.

Adding a preview of the generated confusion matrix for easier reference, though you can build the docs locally and check it out:
orig

Before you begin with code review, there are two points to discuss:

1. Do you agree with the proposed formatting (cc'ing @frankenjoe here)
2. My proposed changes completely remove the option to plot just the percentages. However, I agree that on some cases it might be useful. I could not find a way to do it with a boolean flag though (`percentage=True/False`). I would need to add an extra value, e.g. (`percentage=True/False/'both'`) or something similar. What do you think?

### CODE REVIEW

**Comment by hagenw** [comment]
Thanks for adding this change.
I think when plotting very small confusion matrices it will not work to present both values on them. So I guess in this case it would be nice to three different options to pick from. Unfortunately we decided to use `percentage=True`, which allows only to select from two options. One way forward would be to deprecate the `percentage` argument and replace it with a new argument, or to add another argument and derive what to present from the combination of the new argument together with `percentage`.

**Comment by hagenw** [comment]
And after our internal discussion you would also be in favor of changing the default behavior to `percentage=False` (or present percentage and support information), correct?

**Comment by ATriantafyllopoulos** [comment]
> One way forward would be to deprecate the `percentage` argument and replace it with a new argument, or to add another argument and derive what to present from the combination of the new argument together with `percentage`.

I'm okay with both. Perhaps adding a `support: bool` argument to indicate whether support information should be added (only relevant when `percentage=True` is the best here).

> And after our internal discussion you would also be in favor of changing the default behavior to percentage=False (or present percentage and support information), correct?

No, if we show the support information, I'm okay with showing percentage as well, so we could keep the default behavior as `percentage=True`.

**Comment by frankenjoe** [comment]
But the coloring can only reflect either percentages or absolute numbers. How should we cope with that?

**Comment by hagenw** [comment]
> But the coloring can only reflect either percentages or absolute numbers. How should we cope with that?

We could use the following two options:

* `percentage`
* `show_second`   (ignore the silly name, we need to find a better one)

If `show_second=True` we add the not selected option (percentage or support info) in brackets below and the color is given by the selected option.

So, in the figure @ATriantafyllopoulos is shopwing above the color should be after the support info.

**Comment by ATriantafyllopoulos** [comment]
> If show_second=True we add the not selected option (percentage or support info) in brackets below and the color is given by the selected option.

Although the code is very easy to configure, I would urge for caution here. The reason why this whole conversation started is because I consider the use of simple percentages without support misleading (e.g. in some problem we had the case that one class had only 9 samples, and we were getting only 1 of them correct. The percentage-wise matrix was showing 11% which seemed decent given the task, but of course it was completely unreliable).

Coloring is an important aspect of a confusion matrix as it draws the attention of the user (much like boldface does in a table). I think that the coloring should always be done based on the "primary" information, which should be the number of samples, otherwise it would seem that we are intentionally trying to mislead the reader.

tl;dr: It's trivial for me to add an option in the code if you guys want it, I would just argue against it, but I'll do what you want.

**Comment by hagenw** [comment]
OK, so @ATriantafyllopoulos would suggest the following:

We stay with `percentage` as the only argument, but always color by support info and also show it as first information. Percentag can then only be added as the additional second row.

This would change the way how we currently present the confusion matrix in most of our benchmarks, so I guess this would require a little bit more discussion inside the next research meeting?

**Comment by ATriantafyllopoulos** [comment]
> We stay with percentage as the only argument

I'm also fine with adding a second argument to control the use of support when percentage=True (see here: https://github.com/audeering/audplot/pull/13#issuecomment-884842380)

What I'm arguing in my last comment is that iff support is there, then the coloring should be done exclusively by it (if not, then of course the percentage will be used). So I am against having both percentage and support and coloring by percentage

**Comment by frankenjoe** [comment]
> So I am against having both percentage and support and coloring by percentage

Mhh, but doesn't it actually make to color by the primary format and always show the other format in brackets in the second row? I.e. when `percentage=False` we color by absolute values and show absolute on top and percentage in brackets below that. And if `percentage=True` we color by percentage and show percentage on top and absolute value in brackets below that. This would not change the current behavior and function signature, but simply add a little more information to picture.

**Comment by ATriantafyllopoulos** [comment]
> Mhh, but doesn't it actually make sense to color by the primary format and always show the other format in brackets in the second row?

Yes, this would make sense! And it's easy to implement.

However, I would always choose to show the original confusion matrix without the percentages unless I'm forced to :P But I anyways rarely use `audplot` anyway so I am not the primary customer for this, I prefer to keep control of the plots and do it manually.

So let's do it any way you guys propose, I'll stop protesting :D Perhaps it's best to discuss on Monday what should be the recommended way to officially report confusion matrices moving forward, we implement that as the default, and then we add as much customization as you guys want

**Comment by frankenjoe** [comment]
> However, I would always choose to show the original confusion matrix without the percentages unless I'm forced to

Ok, so you would opt to add an additional argument that controls if the user wants to show the support or not?

**Comment by frankenjoe** [comment]
> But I anyways rarely use `audplot` anyway so I am not the primary customer for this, I prefer to keep control of the plots and do it manually.

The idea of `audplot` is that users do not have to create their plots manually. This avoids we do the same work over and over again. So if you have ideas for missing features please create issues or PRs so that we can extend `audplot` in a direction it covers as many use-cases as possible.

**Comment by ATriantafyllopoulos** [comment]
> The idea of audplot is that users do not have to create their plots manually

Sorry, but I often end up tweaking a lot of things (from labels and color maps in confusion matrices to what I show in scatter plots, where I put the legends, and so on) so I can better match the "feel" of the publication I'm targeting. So I'm personally in need of customization and I'm already used to doing it with seaborn and matplotlib, and I think `audplot` would make my life harder, not easier..

If you guys are aiming to create a standard "audEERING feel" for standardized reporting, I'm clueless as to what it should be (except switching the `cmap` from `Blues` to `Reds` ;) ) and I'll follow whatever guidelines you create and use `audplot` for that (except in cases where I strongly disagree such as showing only percentawise confusion matrices which prompted this PR and discussions)

**Comment by hagenw** [comment]
For further discussion I will summarize our options.

@ATriantafyllopoulos would vote for having:

1. `percentage=False`: showing number of samples and color after that

image

2. `percentage=True`: showing percentage as extra information on the number of samples plots

image

In addition, two other plots are in theory possible

3. Showing only percentage and coloring after that

image

4. Showing number of samples as extra information on the percentage plots

image

**Comment by hagenw** [comment]
There is one problem if we go only for 1. and 2.: the percentage option will not work for very small figures

image

**Comment by ATriantafyllopoulos** [comment]
Added yesterday's agreed-upon changes in 8d21659. I added an additional `support: bool = False` argument which controls the use of support information when `percentage==True`. All in all, we have three options:

* `percentage=False` (default)
default

* `percentage=True, support=False`
percentage

* `percentage=True, support=True`
percentage_support

**Comment by hagenw** [comment]
Sorry, seems we worked on the same stuff. I opened https://github.com/audeering/audplot/pull/15

**Comment by ATriantafyllopoulos** [comment]
> Sorry, seems we worked on the same stuff. I opened #15

oh! I should have known by the suspicious push you did earlier...

How to proceed? Should I undo my commits?

**Comment by hagenw** [comment]
> > Sorry, seems we worked on the same stuff. I opened #15
> 
> oh! I should have known by the suspicious push you did earlier...
> 
> How to proceed? Should I undo my commits?

No, should be fine. I changed my pull request to go into `master` directly.
We just need to decide if we want to proceed with #16 or use the solution presented here. The main difference is that #16 allows for 4 different options of presentations and here we allow for three different options.

**Comment by ATriantafyllopoulos** [comment]
> > > Sorry, seems we worked on the same stuff. I opened #15
> > 
> > 
> > oh! I should have known by the suspicious push you did earlier...
> > How to proceed? Should I undo my commits?
> 
> No, should be fine. I changed my pull request to go into `master` directly.
> We just need to decide if we want to proceed with #16 or use the solution presented here. The main difference is that #16 allows for 4 different options of presentations and here we allow for three different options.

Cool, let's go with yours then!
